src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -902,6 +902,15 @@ public IPointerTypeSymbol CreatePointerTypeSymbol(ITypeSymbol pointedAtType)
 
         protected abstract IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol elementType);
 
+        // PERF: Traces show that analyzers may use this method frequently, often requesting
+        // the same symbol over and over again. In certain instances this can even consume 1% of
+        // the total CPU time of compilation. This is an extremely simple cache that evicts on
+        // hash code conflicts, but seems to do the trick. The size is mostly arbitrary. My guess
+        // is that there are maybe a couple dozen analyzers in the solution and each one has
+        // ~0-2 unique well-known types, and the chance of hash collision is very low.
+        private ConcurrentCache<string, INamedTypeSymbol> _getTypeCache =
+            new ConcurrentCache<string, INamedTypeSymbol>(50);
```

Comment:
but you don't actually pass a comparer here.  which means you just get `EqualityComparer<TKey>.Default`, which should have the same behavior as the old code (except for null handling).

Reply:
Woops! Good news: it's already fast enough to drop it off the perf profile. Better news: I can make it even faster!

Reply:
Note: i think doing normal .Equals here is good.  I'm not sure people will have the exact same reference for the full name each time they call this.  Having this have value-semantics for the name seems sensible to me :)

Reply:
Well, one thing I noticed is that a significant amount of time was spent in just calculating the hash code of the string. I'll check and see if it's still hot in the trace (even if the total time in GetTypeByMetadataName is now minimal).

Reply:
Works for me :)

Reply:
OK, when I use default equality, I get 2 samples (out of 67,000) in the hash function. When I move to reference equality, there are now zero samples in the hash. I think I'm going to go with reference equality. I don't really think that value equality is buying us much, because an analyzer allocating a new string for the type name every invocation seems like it has more significant problems.